src/core/algorithms/ind/faida/candidate_generation/apriori_candidate_generator.cpp

```diff
@@ -0,0 +1,99 @@
+    auto new_right = CombineCCs(*(first.right()), *(second.right()), ccs_on_level);
+
+    int const size = new_left->GetColumnIndices().size();
+
+    if (new_left->GetTableIndex() == new_right->GetTableIndex()) {
+        for (int i : new_left->GetColumnIndices()) {
+            for (int j : new_right->GetColumnIndices()) {
+                if (i == j) return false;
+            }
+        }
+    }
+
+    int const num_checks = size - 2;
+    if (num_checks <= 0) {
+        candidates.emplace_back(std::move(new_left), std::move(new_right));
+        return true;
+    }
+
+    std::vector<int> test_left;
+    std::vector<int> test_right;
+    test_left.reserve(size - 1);
+    test_right.reserve(size - 1);
+
+    for (int check = 0; check < num_checks; check++) {
+        test_left.clear();
+        test_right.clear();
+        for (int col_idx = 0; col_idx < size; col_idx++) {
+            if (col_idx != check) {
+                test_left.push_back(new_left->GetColumnIndices()[col_idx]);
+                test_right.push_back(new_right->GetColumnIndices()[col_idx]);
+            }
+        }
+        SimpleIND const qwe =
+                SimpleIND(std::make_shared<SimpleCC>(new_left->GetTableIndex(), test_left),
+                          std::make_shared<SimpleCC>(new_right->GetTableIndex(), test_right));
+        if (last_result_set.find(&qwe) == last_result_set.end()) {
+            return false;
+        }
+    }
+
+    candidates.emplace_back(std::move(new_left), std::move(new_right));
+    return true;
+}
+
+std::vector<SimpleIND> CreateCombinedCandidates(std::vector<SimpleIND> const& inds) {
+    std::vector<SimpleIND const*> last_result(inds.size());
+    std::transform(inds.begin(), inds.end(), last_result.begin(),
+                   [](SimpleIND const& i) { return &i; });
+    std::sort(last_result.begin(), last_result.end(),
+              [](SimpleIND const* a, SimpleIND const* b) { return *a < *b; });
+
+    std::unordered_set<SimpleIND const*> last_result_set(last_result.begin(), last_result.end());
+
+    std::unordered_set<std::shared_ptr<SimpleCC>> ccs_on_level;
+    std::vector<SimpleIND> candidates;
+    for (auto left_iter = last_result.begin(); left_iter != last_result.end(); left_iter++) {
+        for (auto right_iter = std::next(left_iter); right_iter != last_result.end();
+                                                                                    right_iter++) {
+            auto const& first_ind = **left_iter;
+            auto const& second_ind = **right_iter;
+
+            if (!first_ind.StartsWith(second_ind)) {
+                break;
+            }
+
+            if (first_ind.left()->GetLastColumn() == second_ind.left()->GetLastColumn() ||
+                first_ind.right()->GetLastColumn() == second_ind.right()->GetLastColumn()) {
+                continue;
+            }
+
+            CreateCandidate(first_ind, second_ind, ccs_on_level, last_result_set, candidates);
```

Comment:
nit: you can use `std::reference_wrapper` to get rid of the need to write a transformation and a custom comparator for sorting. There is also no need to write a double dereference (**).
The only drawback I find is that it is not possible to fill `last_result_set` using the default constructor due to the fact that the types are different. I think this version of the function looks a little better:

But it seems to me that this is insignificant and the pros outweigh the cons.
```cpp
    std::vector<std::reference_wrapper<const SimpleIND>> last_result(inds.begin(), inds.end());
    std::sort(last_result.begin(), last_result.end(), std::less<const SimpleIND&>{});

    std::unordered_set<const SimpleIND*> last_result_set;
    for (const auto& ind_rw : last_result) last_result_set.insert(&ind_rw.get());

    std::unordered_set<std::shared_ptr<SimpleCC>> ccs_on_level;
    std::vector<SimpleIND> candidates;
    for (auto left_iter = last_result.begin(); left_iter != last_result.end(); ++left_iter) {
        for (auto right_iter = std::next(left_iter); right_iter != last_result.end();
             ++right_iter) {
            SimpleIND const& left_ind = left_iter->get();
            SimpleIND const& right_ind = right_iter->get();

            if (!left_ind.StartsWith(right_ind)) {
                break;
            }

            if (left_ind.left()->GetLastColumn() == right_ind.left()->GetLastColumn() ||
                left_ind.right()->GetLastColumn() == right_ind.right()->GetLastColumn()) {
                continue;
            }

            CreateCandidate(left_ind, right_ind, ccs_on_level, last_result_set, candidates);
```